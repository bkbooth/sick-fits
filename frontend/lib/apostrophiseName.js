const ESS = 's'

export default function apostrophiseName(name) {
  const nameEndsWithEss = name[name.length - 1].toLowerCase() === ESS
  return `${name}'${nameEndsWithEss ? '' : ESS}`
}
