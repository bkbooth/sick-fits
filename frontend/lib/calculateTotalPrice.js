/**
 * Calculate the total price of a cart by summing cart items' price multiplied by quantity
 *
 * @param {CartItem[]} cart
 * @param
 *
 * @returns {number}
 */
export default function calculateTotalPrice(cart) {
  return cart.reduce((total, cartItem) => {
    if (!cartItem.item) return total;
    return total + cartItem.quantity * cartItem.item.price;
  }, 0);
}

/**
 * @typedef {Object} CartItem
 * @property {number} quantity
 * @property {Object} item
 * @property {number} item.price
 */
