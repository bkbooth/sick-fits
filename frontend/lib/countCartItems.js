/**
 * Count the total number of items in a cart
 *
 * @param {CartItem[]} cart
 *
 * @returns {number}
 */
export default function countCartItems(cart) {
  return cart.reduce((total, cartItem) => total + cartItem.quantity, 0);
}

/**
 * @typedef {Object} CartItem
 * @property {number} quantity
 */
