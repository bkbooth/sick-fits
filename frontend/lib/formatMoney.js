export default function formatMoney(amount) {
  return new Intl.NumberFormat('en-AU', {
    style: 'currency',
    currency: 'AUD',
    // if its a whole, dollar amount, leave off the .00
    minimumFractionDigits: amount % 100 === 0 ? 0 : 2,
  }).format(amount / 100);
}
