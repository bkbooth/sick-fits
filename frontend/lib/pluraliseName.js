const ESS = 's'

export default function pluraliseName(name) {
  const nameEndsWithEss = name[name.length - 1].toLowerCase() === ESS
  return `${name}'${nameEndsWithEss ? '' : ESS}`
}
