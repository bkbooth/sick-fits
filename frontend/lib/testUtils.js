import casual from 'casual';

// seed it so we get consistent results
casual.seed(777);

export const fakeUser = () => ({
  __typename: 'User',
  id: casual.uuid,
  name: casual.name,
  email: casual.email,
  permissions: ['ADMIN'],
  orders: [],
  cart: [],
});

export const fakeItem = () => ({
  __typename: 'Item',
  id: casual.uuid,
  title: casual.title,
  description: casual.description,
  price: casual.integer(1000, 30000),
  image: `${casual.word}.jpg`,
  largeImage: `${casual.word}-large.jpg`,
  user: null,
});

export const fakeCartItem = overrides => ({
  __typename: 'CartItem',
  id: casual.uuid,
  quantity: casual.integer(1, 5),
  item: fakeItem(),
  user: fakeUser(),
  ...overrides,
});

export const fakeOrderItem = () => ({
  __typename: 'OrderItem',
  id: casual.uuid,
  title: casual.title,
  description: casual.description,
  price: casual.integer(1000, 30000),
  image: `${casual.word}.jpg`,
  quantity: casual.integer(1, 5),
});

export const fakeOrder = () => ({
  __typename: 'Order',
  id: casual.uuid,
  charge: `ch_${casual.uuid.split('-')[4]}`,
  total: casual.integer(10000, 100000),
  items: [fakeOrderItem(), fakeOrderItem()],
  createdAt: casual.date('YYYY-MM-DD[T]HH:mm:ss[Z]'),
  user: fakeUser(),
});

export function typeInto(wrapper, { inputType = 'input', name, type = 'text', value }) {
  wrapper
    .find(`${inputType}[name="${name}"]`)
    .simulate('change', { target: { name, type, value } });
}
