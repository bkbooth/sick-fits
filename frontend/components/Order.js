import React from 'react'
import PropTypes from 'prop-types'
import Head from 'next/head'
import { Query } from 'react-apollo'
import gql from 'graphql-tag'
import { format } from 'date-fns'
import Error from './ErrorMessage'
import OrderStyles from './styles/OrderStyles'
import formatMoney from '../lib/formatMoney'

const SINGLE_ORDER_QUERY = gql`
  query SINGLE_ORDER_QUERY($id: ID!) {
    order(id: $id) {
      id
      charge
      total
      createdAt
      user {
        id
      }
      items {
        id
        title
        description
        price
        image
        quantity
      }
    }
  }
`

class Order extends React.Component {
  static propTypes = {
    id: PropTypes.string.isRequired,
  }

  render() {
    return <Query query={SINGLE_ORDER_QUERY} variables={{ id: this.props.id }}>
      {({ data, error, loading }) => {
        if (error) return <Error error={error} />
        if (loading) return <p>Loading...</p>
        const { order } = data
        return (
          <OrderStyles>
            <Head>
              <title>Sick Fits! - Order {order.id}</title>
            </Head>
            <p>
              <span>Order ID</span>
              <span>{order.id}</span>
            </p>
            <p>
              <span>Charge</span>
              <span>{order.charge}</span>
            </p>
            <p>
              <span>Date</span>
              <span>{format(order.createdAt, 'MMMM d, yyyy h:mm a')}</span>
            </p>
            <p>
              <span>Order Total</span>
              <span>{formatMoney(order.total)}</span>
            </p>
            <p>
              <span>Items Count</span>
              <span>{order.items.length}</span>
            </p>
            <div className="items">
              {order.items.map(item => (
                <div className="order-item" key={item.id}>
                  <img src={item.image} alt={item.title} />
                  <div className="item-details">
                    <h2>{item.title}</h2>
                    <p>Qty: {item.quantity}</p>
                    <p>Each: {formatMoney(item.price)}</p>
                    <p>Sub-total: {formatMoney(item.price * item.quantity)}</p>
                    <p>{item.description}</p>
                  </div>
                </div>
              ))}
            </div>
          </OrderStyles>
        )
      }}
    </Query>
  }
}

export default Order
