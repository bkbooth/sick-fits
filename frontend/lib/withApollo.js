import ApolloClient, { InMemoryCache } from 'apollo-boost';
import withApollo from 'next-with-apollo';
import { LOCAL_STATE_QUERY } from 'components/Cart';
import { GRAPHQL_URL } from '../config';

function createApolloClient({ headers, initialState }) {
  return new ApolloClient({
    uri: GRAPHQL_URL,
    cache: new InMemoryCache().restore(initialState || {}),
    request: operation => {
      operation.setContext({
        fetchOptions: {
          credentials: 'include',
        },
        headers,
      });
    },
    clientState: {
      resolvers: {
        Mutation: {
          toggleCart(_parent, _args, { cache }, _info) {
            const { cartOpen } = cache.readQuery({ query: LOCAL_STATE_QUERY });
            const result = { data: { cartOpen: !cartOpen } };
            cache.writeData(result);
            return result;
          },
        },
      },
      defaults: {
        cartOpen: false,
      },
    },
  });
}

export default withApollo(createApolloClient);
