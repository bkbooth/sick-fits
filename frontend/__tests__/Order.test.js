import { MockedProvider, wait } from '@apollo/react-testing';
import { mount } from 'enzyme';
import { act } from 'react-dom/test-utils';
import { fakeOrder } from 'lib/testUtils';
import Order, { SINGLE_ORDER_QUERY } from 'components/Order';

async function render(mocks, orderId) {
  const wrapper = mount(
    <MockedProvider mocks={mocks}>
      <Order orderId={orderId} />
    </MockedProvider>
  );
  await act(() => wait());
  wrapper.update();
  return wrapper;
}

describe('<Order />', () => {
  it('renders and matches snapshot', async () => {
    const order = fakeOrder();
    const mocks = [
      {
        request: { query: SINGLE_ORDER_QUERY, variables: { orderId: order.id } },
        result: { data: { order } },
      },
    ];
    const wrapper = await render(mocks, order.id);
    expect(wrapper.find('div[data-test="order"]')).toMatchSnapshot();
  });

  it('errors with a not found order', async () => {
    const mocks = [
      {
        request: { query: SINGLE_ORDER_QUERY, variables: { orderId: 'invalid-id' } },
        error: new Error('Order not found'),
      },
    ];
    const wrapper = await render(mocks, 'invalid-id');
    expect(wrapper.find('[data-test="graphql-error"]')).toMatchSnapshot();
  });
});
