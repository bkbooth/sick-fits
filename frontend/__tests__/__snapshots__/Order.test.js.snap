// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Order /> errors with a not found order 1`] = `
<p
  data-test="graphql-error"
>
  <strong>
    Shoot!
  </strong>
  Network error: Order not found
</p>
`;

exports[`<Order /> renders and matches snapshot 1`] = `
<div
  className="OrderStyles-sc-4oqalm-0 eDzsVm"
  data-test="order"
>
  <Head>
    <Component
      handleStateChange={null}
      inAmpMode={false}
      reduceComponentsToState={[Function]}
    />
  </Head>
  <p>
    <span>
      Order ID
    </span>
    <span>
      3a430a73-c9e9-4bcc-b46a-41618965ffea
    </span>
  </p>
  <p>
    <span>
      Charge
    </span>
    <span>
      ch_f090492fd03e
    </span>
  </p>
  <p>
    <span>
      Date
    </span>
    <span>
      27th April, 2016 10:51 PM
    </span>
  </p>
  <p>
    <span>
      Total
    </span>
    <span>
      $565.17
       for 
      2
       item
      s
    </span>
  </p>
  <div
    className="items"
  >
    <div
      className="order-item"
      key="18d9d810-88de-4ec8-884c-84bf78c1398f"
    >
      <img
        alt="Debitis rem similique"
        src="aspernatur.jpg"
      />
      <div
        className="item-details"
      >
        <h2>
          Debitis rem similique
        </h2>
        <p>
          Quantity: 
          1
          <br />
          Price: 
          $14.74
          <br />
          Subtotal: 
          $14.74
        </p>
        <p>
          Sed sapiente soluta corporis commodi. Nisi cum ducimus. Doloremque maiores esse atque et. Explicabo dolores ipsam laboriosam non itaque voluptas ea ducimus autem.
        </p>
      </div>
    </div>
    <div
      className="order-item"
      key="7bcd266d-b805-4ab0-80e0-4cef4042cadc"
    >
      <img
        alt="Qui enim et"
        src="et.jpg"
      />
      <div
        className="item-details"
      >
        <h2>
          Qui enim et
        </h2>
        <p>
          Quantity: 
          4
          <br />
          Price: 
          $178.90
          <br />
          Subtotal: 
          $715.60
        </p>
        <p>
          In sunt occaecati nulla aut sunt. Rerum at sint dolores fugiat quis et voluptas.
        </p>
      </div>
    </div>
  </div>
</div>
`;
